nevergrad/functions/multiobjective.py

```diff
@@ -0,0 +1,387 @@
+#    License along with DEAP. If not, see <http://www.gnu.org/licenses/>.
+#
+#    Modified by O. Teytaud as needed for Nevergrad.
+
+from ..instrumentation import InstrumentedFunction
+from math import log, floor
+import random
+from typing import (Tuple, Any, Callable, List, Optional, Dict, ValuesView, Iterator,
+                    TypeVar, Generic, Deque, Iterable)
+import warnings
+import numpy
+#from ..optimization import base
+from ..optimization.base import Candidate
+from ..common.typetools import ArrayLike
+from ..instrumentation import InstrumentedFunction
+
+def hypervolume(pointset, ref):
+    """Compute the absolute hypervolume of a *pointset* according to the
+    reference point *ref*.
+    """
+    hv = _HyperVolume(ref)
+    return hv.compute(pointset)
+
+
+class MultiobjectiveFunction(InstrumentedFunction):
+    """Given several functions, and threshold on their values (above which solutions are pointless),
+    this function returns a single-objective function, correctly instrumented, the minimization of which
+    yields a solution to the original multiobjective problem.
+    
+    multi_objective_function: objective functions, to be minimized, of the original multiobjective problem.
+    upper_bounds: upper_bounds[i] is a threshold above which x is pointless if functions[i](x) > upper_bounds[i].
+    
+    Returns an objective function to be minimized (it is a single objective function).
+    Warning: this function is not stationary.
+    The minimum value obtained for this objective function is -h,
+    where h is the hypervolume of the Pareto front obtained, given upper_bounds as a reference point.
+    """
+
+
+    def __init__(self, multiobjective_function: Callable[..., Tuple[float, ...]], upper_bounds: Tuple[float, ...]) -> None:
+        self.multi_objective_function = multiobjective_function
+        self.upper_bounds = numpy.array(upper_bounds)
+        self.pointset: Dict[Candidate, ArrayLike] = {}
+        self.best_hypervolume = -float("Inf")
+
+    def compute_aggregate_loss(self, x: Candidate, v: Tuple[float]) -> None:
+        # We compute the hypervolume
+        my_hypervolume = hypervolume(list(self.pointset.values()) + [v], self.upper_bounds)
+        #print(x, " --> ", v, "     hv=", my_hypervolume)
+        if my_hypervolume > self.best_hypervolume:  # This point is good! Let us give him a great mono-fitness value.
+            self.best_hypervolume = my_hypervolume
+            if tuple(x) in self.pointset:
+                assert v == self.pointset[tuple(x)]  # We work noise-free...
+            self.pointset[tuple(x)] = v
+            return -my_hypervolume
+
+        # Now we compute for each axis 
+        distance_to_pareto = float("Inf")
+        for y in self.pointset.values():
+            if all(y <= v):
+                distance_to_pareto = min(distance_to_pareto, min(v-y))
+        return -my_hypervolume + distance_to_pareto
+
+    def __call__(self, candidate: Candidate):
```

Comment:
to make it more transparent, this (and other functions) should not take Candidate as input, just *args and **kwargs